velox/exec/TreeOfLosers.h

```diff
@@ -15,103 +15,206 @@
  */
 #pragma once
 
+#include <algorithm>
 #include <memory>
 #include <optional>
 #include <vector>
 
+#include "velox/common/base/Exceptions.h"
+#include "velox/common/base/SimdUtil.h"
+
+#include <folly/Likely.h>
+
 namespace facebook::velox {
 
-template <typename Value, typename Source>
+// Abstract class defining the interface for a stream of values to be
+// merged by TreeOfLosers or MergeArray. In addition to these, the
+// MergeStream must have a way of accessing its first value and
+// popping off the first value. TreeOfLosers and similar do not call
+// these, so these are left out of this interface.
+class MergeStream {
+  // True if this has a value. If this returns true, it is valid to
+  // call <. A false value means that there will not be any more data
+  // in 'his'.
+  virtual bool hasData() const = 0;
+
+  // Returns true if the first element of 'this' is less than the first element
+  // of 'other'. hasData() must be true of 'this' and 'other'.
+  virtual bool operator<(const MergeStream& other) const = 0;
+};
+
+// Implements a tree of loserrs algorithm for merging ordered
+// streams. The TreeOfLosers owns two or more instances of
+// Source. At each call of next(), it returns the Source that has
```

Comment:
The names Source and MergeStream are not consistent. Perhaps, rename to Source and MergeSource or Stream and MergeStream.